### PR TITLE
Implement FullSyncIterDataPipe

### PR DESCRIPTION
**Pull request by ejguan**
### Changes

- Add `_PrefetchExecutor` to run prefetching in multi-threading
  - This should be reused by `PrefetchIterDataPipe`
- Add `FullSyncIterDataPipe`
  - [x] Tested the result using elastic training and normal distributed training.
  - [x] Add distributed unit test for GLOO
  - [x] Add serialization test (How to do checkpoint for `FullSyncIterDataPipe` is unclear to me)
    - Serialization is working properly as I am using spawn to serialize/deserialize such DataPipe to distributed processes.

### CODE REVIEW

File: torchdata/datapipes/iter/util/prefetch.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    Expected data provided to callback function in ``_PrefetchExecutor``.
+    """
+    index: int
+    error: Optional[BaseException] = None
+
+    def has_error(self) -> bool:
+        return self.error is not None
+
+
+class _PrefetchExecutor:
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/prefetch.py:49]: This executor takes reference from the internal implementation: https://fburl.com/code/7dk6mvs4
On the top of the implementation, I added `prefetch_size` to make sure it can work with `Prefetch` in the future.

File: test/test_distributed.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    )
+
+
+class DistributedTest(TestCase):
+    @staticmethod
+    def _test_fullsync(rank, world_size, backend):
+        dist.init_process_group(backend, rank=rank, world_size=world_size)
+        # Use a prime number to make sure uneven data sharding
+        data_length = 23
+        dp = IterableWrapper(list(range(data_length))).sharding_filter().fullsync()
```
Comment (ejguan) [inline test/test_distributed.py:53]: Without `fullsync`, this pipeline would hang forever.

File: torchdata/_constants.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Use the same timeout as PyTorch Distributed
+default_timeout_in_s = 30 * 60
```
Comment (NivekT) [inline torchdata/_constants.py:8]: Just a comment (no change) - we should put other things such as default buffer size here too.

File: torchdata/datapipes/iter/util/prefetch.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+
+@functional_datapipe("fullsync")
+class FullSyncIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    Synchronizes data across distributed processes to prevent hanging during training,
+    which is caused by uneven sharded data (functional name: ``fullsync``). It should
+    be appended at the end of the graph of ``DataPipe`` by ``DistributedReadingService``
+    automatically.
+
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/prefetch.py:115]: Question: do we recommend against usage of this DataPipe outside of a ReadingService? If not, can we potentially include an example?

File: torchdata/datapipes/iter/util/prefetch.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    def __init__(self, datapipe: IterDataPipe, timeout=default_timeout_in_s):
+        self.datapipe = datapipe
+        self.timeout = timeout
+
+        self._process_group = None
+        self._world_size = 1
+
+        self._lock = threading.RLock()
+        self._cv = threading.Condition(lock=self._lock)
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/prefetch.py:129]: super nit: can we rename this to `_condition_variable`? Feel free to ignore.

File: torchdata/datapipes/iter/util/prefetch.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            with self._lock:
+                self._idx += 1
+
+    def fetch_next(self):
+        return next(self.datapipe_iterator)
+
+    def _done_callback_fn(self, index: int, f: Future):
+        if f.exception():
+            with self._lock:
+                self._end_flag = True
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/prefetch.py:83]: Question: for any exception (including `StopIteration`), is this setting the `_end_flag = True` but is not re-raising the exception? It will then be handled by `_executor` and `callback_fn` if available?

File: torchdata/datapipes/iter/util/prefetch.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    def _callback_fn(self, exp: Expected) -> None:
+        with self._cv:
+            if exp.has_error():
+                if not isinstance(exp.error, StopIteration):
+                    self._error = exp.error  # type: ignore[assignment]
+                self._sync_counter = torch.tensor([0], dtype=torch.int32)
+            else:
+                self._sync_counter = torch.tensor([1], dtype=torch.int32)
+            dist.all_reduce(
+                tensor=self._sync_counter,
```
Comment (NivekT) [inline torchdata/datapipes/iter/util/prefetch.py:146]: Question: what is `_sync_counter` doing here?

File: torchdata/datapipes/iter/util/prefetch.py (hunk 4 above)
Comment (ejguan) [reply torchdata/datapipes/iter/util/prefetch.py:115]: Makes sense. Will add it even though we should always recommend users relying on RS

File: torchdata/datapipes/iter/util/prefetch.py (hunk 5 above)
Comment (ejguan) [reply torchdata/datapipes/iter/util/prefetch.py:129]: I would like keeping it as it is. cv is used because I take the reference from Python official example

https://docs.python.org/3/library/threading.html#condition-objects

File: torchdata/datapipes/iter/util/prefetch.py (hunk 6 above)
Comment (ejguan) [reply torchdata/datapipes/iter/util/prefetch.py:83]: Yeah. It's callback in a separate thread to set flag to DataPipe. And exception of StopIteration should be re-raised when __iter__ is invoked from Datapipe

File: torchdata/datapipes/iter/util/prefetch.py (hunk 7 above)
Comment (ejguan) [reply torchdata/datapipes/iter/util/prefetch.py:146]: It's used to count the number of distributed processes that has fetched one data.
When one process ends or raises Error, the counter would be less than world size. Then the full sync should stop the iteration to prevent any uneven data sent to model training.

Comment (NivekT) [reply torchdata/datapipes/iter/util/prefetch.py:146]: Got it. IIUC, this is truncating rather than padding?

Comment (ejguan) [reply torchdata/datapipes/iter/util/prefetch.py:146]: Good question. I will add padding like feature to the tracker with a lower priority. There are many ways to do padding. Needs some careful discussion afterwards.

File: torchdata/datapipes/iter/__init__.pyi.in [hunk 8]
```diff
@@ -8,6 +8,7 @@ ${init_base}
 # classes/objects here, even though we are not injecting extra code into them at the moment.
 
 from .util.decompressor import CompressionType
+from torchdata._constants import default_timeout_in_s
```
Comment (ejguan) [inline torchdata/datapipes/iter/__init__.pyi.in:11]: Here

File: torchdata/datapipes/iter/util/prefetch.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            self._done_callback = False
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+        state = (
+            self.datapipe,
+            self.timeout,
+        )
+        return state
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/prefetch.py:212]: IMHO, checkpoint for fullsync or prefetch is a little tricky.
Let's confirm the expected behavior. When we do checkpoint, we should pause any further prefetching and save all prefetched data into a buffer. Then, we serialize the buffer ant inner datapipe (because we have to serialize datapipe after prefetching is done). And, only when we start iteration again, would we start prefetching again.

WDYT: @VitalyFedyunin @NivekT 

Then, the whole logic of fullsync should be changed. This is even more complicated when the data ends when put the prefetched data into the buffer. I might open a new PR to achieve serialization.

Comment (Miiira) [reply torchdata/datapipes/iter/util/prefetch.py:212]: Yea I think we should stop the prefetch and capture current data. I feel this can be similar to internal client snapshot, so https://fburl.com/code/6hrjawgh may be helpful for reference

File: torchdata/datapipes/iter/util/prefetch.py (hunk 7 above)
Comment (Miiira) [reply torchdata/datapipes/iter/util/prefetch.py:160]: We have some users doing padding/exhaustive reading like https://fburl.com/code/4qk4oks1

**Review by Miiira — APPROVED** [review]

File: torchdata/datapipes/iter/util/prefetch.py (hunk 7 above)
Comment (ejguan) [reply torchdata/datapipes/iter/util/prefetch.py:160]: Sounds great. I will add this to one of TODOs other than WrapAround.

**Comment by ejguan** [comment]
I will land this PR for now. List two follow-up works:
- [ ] Implement serialization logic for fullsync
- [ ] Implement padding for fullsync (I need to take a deeper look to see if we can combine it with wraparound)

**Review by NivekT — APPROVED** [review]
